### PR TITLE
world, biome generation

### PR DESCRIPTION
**Pull request by screret**

### CODE REVIEW

File: crates/ocg_common/src/voxel/mod.rs [hunk 1]
```diff
@@ -1,3 +1,6 @@
 //! Voxel world management
+#![allow(missing_docs)]
```
Comment (eigenraven) [inline crates/ocg_common/src/voxel/mod.rs:2]: Can you remove this and add the missing doc comments?

File: crates/ocg_common/Cargo.toml [hunk 2]
```diff
@@ -14,3 +14,12 @@ ocg_schemas.workspace = true
...
 anyhow.workspace = true
 bevy.workspace = true
+bevy_math.workspace = true
+noise.workspace = true
+rand.workspace = true
+rand_xoshiro.workspace = true
+thread_local.workspace = true
+lru.workspace = true
+lazy_static.workspace = true
+serde.workspace = true
```
Comment (eigenraven) [inline crates/ocg_common/Cargo.toml:24]: Why do you need serde here?

File: lib/ocg_schemas/Cargo.toml [hunk 3]
```diff
@@ -23,6 +23,7 @@ bytemuck.workspace = true
 bitflags.workspace = true
 bitvec.workspace = true
 bevy_math.workspace = true
+bevy.workspace = true
```
Comment (eigenraven) [inline lib/ocg_schemas/Cargo.toml:26]: bevy should definitely not go into the schemas library, it's supposed to be for data format manipulation only

File: lib/ocg_schemas/Cargo.toml [hunk 4]
```diff
@@ -33,6 +34,12 @@ rgb.workspace = true
 serde.workspace = true
 smallvec.workspace = true
 thiserror.workspace = true
+rand.workspace = true
```
Comment (eigenraven) [inline lib/ocg_schemas/Cargo.toml:37]: Most of the rand stuff, probably lazy_static and image shouldn't be here either

File: lib/ocg_schemas/src/coordinates.rs [hunk 5]
```diff
@@ -25,6 +25,13 @@ pub const CHUNK_DIM3Z: usize = (CHUNK_DIM * CHUNK_DIM * CHUNK_DIM) as usize;
 /// Chunk dimensions in blocks as a [IVec3] for convenience
 pub const CHUNK_DIM3V: IVec3 = IVec3::splat(CHUNK_DIM);
 
+/// Length of a side of a region in blocks
+pub const REGION_SIZE: i32 = 4 * CHUNK_DIM;
```
Comment (eigenraven) [inline lib/ocg_schemas/src/coordinates.rs:29]: Define region :concern:

File: lib/ocg_schemas/src/coordinates.rs [hunk 6]
```diff
@@ -341,6 +419,13 @@ impl From<AbsBlockPos> for AbsChunkPos {
 // === RelChunkPos
 impl_simple_ivec3_newtype!(RelChunkPos);
 impl_rel_abs_pair!(RelChunkPos, AbsChunkPos);
+
+impl From<AbsChunkPos> for RelChunkPos {
+    fn from(value: AbsChunkPos) -> Self {
+        Self(value.0)
```
Comment (eigenraven) [inline lib/ocg_schemas/src/coordinates.rs:425]: This is wrong, absolute to relative should be done by subtracting an absolute reference point

File: lib/ocg_schemas/src/registry.rs [hunk 7]
```diff
@@ -140,13 +141,13 @@ impl<'a> Display for RegistryNameRef<'a> {
...
 /// Needs to be implemented on any object that can be a part of a Registry
-pub trait RegistryObject {
+pub trait RegistryObject: PartialEq {
     /// Should be trivial
     fn registry_name(&self) -> RegistryNameRef;
 }
 
 /// A data structure for keeping track of a stable mapping between: namespaced strings, numerical IDs and objects.
-#[derive(Serialize, Deserialize)]
+#[derive(Serialize, Deserialize, Resource)]
```
Comment (eigenraven) [inline lib/ocg_schemas/src/registry.rs:150]: These shouldn't be resources directly, there will be proxy objects in common that will have net sync code etc.

File: lib/ocg_schemas/src/voxel/generation/blur.rs [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        let slice_second_last = &self.slice[last - 1];
+        Some([slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), slice_second_last.clone(), self.slice[last].clone()])
+    }
+}
+
+/// do a convolution blur of the data.
+pub fn blur_biomes(input: &[&[SmallVec<[BiomeEntry; EXPECTED_BIOME_COUNT]>]]) -> Vec<Vec<SmallVec<[BiomeEntry; EXPECTED_BIOME_COUNT]>>> {
+    let mut output: Vec<Vec<SmallVec<[BiomeEntry; EXPECTED_BIOME_COUNT]>>> = input.iter().cloned().map(|v| v.to_vec()).collect_vec();
+    for (in_rows, out_row) in Convolution16::new(input).zip(&mut output) {
+        for ((((((((((((((((r0, r1), r2), r3), r4), r5), r6), r7), r8), r9), r10), r11), r12), r13), r14), r15), out) in
```
Comment (eigenraven) [inline lib/ocg_schemas/src/voxel/generation/blur.rs:115]: This is very concerning, can't you make this an array

Comment (screret) [reply lib/ocg_schemas/src/voxel/generation/blur.rs:115]: idk how lol

**Review by eigenraven — APPROVED** [review]

**Pull request merged**